File: solutions/rust/src/1326/greedy.rs
impl Solution {
  pub fn min_taps(n: i32, ranges: Vec<i32>) -> i32 {
      let mut range = ranges
          .iter()
          .enumerate()
          .map(|(i, &r)| (i as i32 - r, i as i32 + r))
          .collect::<Vec<_>>();

      range.sort();
      // 贪心

      // println!("{range:?}");
      let mut cnt = 0;
      let mut cur_left = 0; //下一个要达到的最小left

      let mut max_right = -1;

      let mut i = 0;
      while i <= n && cur_left < n {
          // 找到左侧能满足条件的右侧最大的水龙头
          let mut flag = false;
          while i <= n {
              let (left, right) = range[i as usize];
              if left > cur_left {
                  break;
              }
              flag = true;
              max_right = max_right.max(right);
              i += 1;
          }
          if !flag || (i > n && max_right < n) {
              return -1;
          }
          cnt += 1;
          cur_left = max_right;
      }

      return cnt;
  }
}

File: solutions/cpp/src/1053/greedy.cpp
class Solution
{
public:
  vector<int> prevPermOpt1(vector<int> &arr)
  {
    int n = arr.size();
    for (int left = n - 2; left >= 0; left--)
    {
      // 从右往左找到第一个合适的left位置
      // 例如:[...,3,1] 中3可以被替换，选择left为0
      //    而[...,1,3] 中不存在合适的left位置
      if (arr[left] > arr[left + 1])
      {
        // 从右往左找小于left位置的第一个元素，
        // 若存在相邻重复元素，答案为最左侧的
        int right = n - 1;
        while (arr[right] >= arr[left] || arr[right] == arr[right - 1])
        {
          right -= 1;
        }
        swap(arr[left], arr[right]);
        return arr;
      }
    }

    return arr;
  }
};
